### PR TITLE
Generalising Head Interface to make it Chain Agnostic

### PR DESCRIPTION
**Pull request by ExtremelySunnyYK**
This pr is to accomplish the following tasks:
1. Move Head from txmgr/type to common/type 
    - Chain Agnostic Head is used more than just for Transaction Manager
2. Make Head Chain Agnostic by removing the reliance on common/ethereum and using a Generic Hash Type

Tickets:
Generalising Head Interface:
https://smartcontract-it.atlassian.net/browse/BCI-1291

Moving Head folder location:
https://smartcontract-it.atlassian.net/browse/BCI-1265

### CODE REVIEW

File: common/types/hashable.go [hunk 1]
```diff
@@ -9,4 +9,5 @@ type Hashable interface {
 	comparable
 
 	Bytes() []byte
+	Hex() string
```
Comment (aalu1418) [inline common/types/hashable.go:12]: why is this needed?

for `common.Address` `Hex()` and `String()` do the same thing - and hex encoding is a bit more of an EVM concept

Comment (ExtremelySunnyYK) [reply common/types/hashable.go]: Thanks for pointing that out. My Mistake, as i thought that `Hex()` encoding is a generalised concept. 

I added `Hex()` because it was being called by the logger in `eth_confirmer.go`.

However it looks like `String()` for common.hash also calls `Hex()`, so it makes sense for us to use `String()`. 

Made the following changes:
1. Removed the `Hex()` from the interface. 
2. Used `String()` instead of `Hex()` for `eth_confirmer.go` log.

File: core/chains/evm/txmgr/eth_confirmer.go [hunk 2]
```diff
@@ -1071,7 +1071,7 @@ func (ec *EthConfirmer[CHAIN_ID, HEAD, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE, A
 	ec.lggr.Infow(fmt.Sprintf("Re-org detected. Rebroadcasting transaction %s which may have been re-org'd out of the main chain", attempt.Hash.String()),
 		"txhash", attempt.Hash.String(),
 		"currentBlockNum", head.BlockNumber(),
-		"currentBlockHash", head.BlockHash().Hex(),
+		"currentBlockHash", head.BlockHash().String(),
```
Comment (ExtremelySunnyYK) [inline core/chains/evm/txmgr/eth_confirmer.go:1074]: String() for common.hash also calls Hex(), so it makes sense for us to use String().

**Review by aalu1418 — APPROVED** [review]

**Review by jmank88 — APPROVED** [review]

File: common/types/head.go [hunk 3]
```diff
@@ -1,28 +1,26 @@
 package types
 
-import "github.com/ethereum/go-ethereum/common"
-
 // Head provides access to a chain's head, as needed by the TxManager.
 // This is a generic interface which ALL chains will implement.
 //
 //go:generate mockery --quiet --name Head --output ./mocks/ --case=underscore
-type Head interface {
+type Head[HASH Hashable] interface {
```
Comment (prashantkumar1982) [inline common/types/head.go:7]: HASH should instead be BLOCK_HASH

File: common/txmgr/types/fee_estimator.go [hunk 4]
```diff
@@ -32,9 +32,9 @@ type PriorAttempt[F Fee, TX_HASH types.Hashable] interface {
 // FeeEstimator provides a generic interface for fee estimation
 //
 //go:generate mockery --quiet --name FeeEstimator --output ./mocks/ --case=underscore
-type FeeEstimator[H Head, F Fee, MAXPRICE any, TX_HASH types.Hashable] interface {
+type FeeEstimator[H types.Head[TX_HASH], F Fee, MAXPRICE any, TX_HASH types.Hashable] interface {
```
Comment (prashantkumar1982) [inline common/txmgr/types/fee_estimator.go:35]: All of the TX_HASH in this PR I think should be BLOCK_HASH

File: common/types/head.go (hunk 3 above)
Comment (ExtremelySunnyYK) [reply common/types/head.go:7]: Thanks for the comments Prashant! Is there a reason why it should be BLOCK_HASH instead? Is it because of naming conventions or functional reasons?
